### ISSUE TITLE
Map an index to other paths in Firebase

### ISSUE THREAD

**Issue by danielmahal** [unfold/react-firebase#35]
I would like to map an index of keys to objects in another location in my database. This is valueable in a denormalized structure like this:

```json
{
  "game": {
    "1": {
      "players": {
        "a": true,
        "b": true
      }
    }
  },

  "users": {
    "a": {
      "name": "Daniel"
    },
    "b": {
      "name": "Simen"
    },
    "c": {
      "name": "Seb"
    }
  }
}
```

On a `Game` component I’m subscribing to the index object `game/1/players`, and I want to somehow map this index to the actual `users/x` object.

```js
@connect(props => {
  players: `game/${props.id}/players`
})

class Game extends React.Component {
  render () {
    console.log(this.props.players)
    // Result: { "a": true, "b": true }
    // I want: { "a": "Daniel", "b": "Simen" }
  }
}
```

Currently I would have to do something like this:

```js
import { transform, set, forEach } from 'lodash'

// Get the players index
@connect(props => {
  players: `game/${props.id}/players`
})

// Map index to { "players.a": "users/a" }
@connect(props => {
  ...transform(props.players, (result, value, key) => ({
    ...result,
    [`players.${key}`]: `users/${key}`
  }))
})

// Map { "players.a": "Daniel" } to { "players": { "a": "Daniel" } }
@mapProps(props => {
  const result = { ...props }

  forEach(props, (value, key) => {
    if (/^players\./).test(key) {
      set(result, key, value)
    }
  })

  return result
})
```

Ugh! I don’t really know if this is really working. And maybe there’s a better way?

Would be nice if either:
a) we find a better solution to do this outside react-firebase
b) have some utility inside or outside react-firebase that handles this
c) bake it into the react-firebase api

Any ideas?

**Comment by simenbrekken** [comment]
The reason we removed this capability from the previous major version was because we added support for [declarative queries](https://github.com/unfold/react-firebase#pass-todos-completedtodos-a-function-that-completes-a-todo-completetodo-and-one-that-logs-in-as-props) where differentiating between a subscription and a query [is done by checking if the value is a string (path) or an object](https://github.com/unfold/react-firebase/blob/master/src/connect.js#L25).

@einarlove however came up with an elegant suggestion to this problem where we regain capability to do nested queries while also keeping, and potentially improving declarative query support:

```js
const createQuery = query => ({
  [REACT_FIREBASE_QUERY_SYMBOL]: true,
  query,
})

export default connect(props => ({
  pathSubscription: 'path/to/data',
  deepSubscription: {
    a: {
      aa: 'path/to/data/a/aa',
    }
    b: 'path/to/data/b',
  },
  querySubscription: createQuery({
    path: 'path/to/other/data',
    orderByChild: 'order',
  }),
}))(App)
```

**Comment by einarlove** [comment]
```js
compose(
  connect(props => ({
    game: `games/${props.id}`,
  }),
  connect(props => ({
    players: mapValues(props.game.players, id => `players/${id}`),
  }),
)(Game)
```

**Comment by simenbrekken** [comment]
- [ ] Bring back [`flattenObject` and `expandObject`](https://gist.github.com/simenbrekken/cee77159f0212bf49c75545fab294cd3)
- [ ] Add `createQuery`
- [ ] Update `computeSubscriptions`
- [ ] Add tests

**Comment by danielmahal** [comment]
Ok, because the declarative query is essentially creating a firebase query?

The suggested solution sounds great by the way!

**Comment by simenbrekken** [comment]
Yeah, the declarative query is in essence:

```js
props => ({
  path: '/users',
  orderByChild: 'age',
  limitToFirst: props.limit || 10,
})
```

Is translated to `ref.child('/users').orderByChild('age').limitToFirst(10)`

**Comment by simenbrekken** [comment]
Forgot to close this after it was implemented in #6
